packages/compartment-mapper/NEWS.md

```diff
@@ -1,5 +1,28 @@
 User-visible changes to the compartment mapper:
 
+# Next release
+
+- Adds a support for consistent hashing (SHA-512) of applications:
+  - `nodeReadPowers(fs, crypto)` produces the necessary capabilities for
+    hashing, when passed the Node.js `crypto` module.
+  - `writeArchive` and `makeArchive` accept a `computeSha512` capability and
+    use it to populate the `compartment-map.json` included within the archive
+    with the SHA-512 of every module in the archive.  This ensures that the
+    hash of `compartment-map.json` in a pair of archives is consistent only if
+    every file is consistent.
+  - `importArchive`, `loadArchive`, and `parseArchive` all optionally accept a
+    `computeSha512` capability, use it to verify the integrity of the archive
+    and provide the hash of the contained `compartment-map.json` so the caller
```

Comment:
Two concerns:
* Do the load/execute functions *return* the computed hash? I think that's too late: the wrong code has already had a chance to execute with whatever authorities you passed in. A safer API would be to pass the expected hash *in*, have the `compartment-map.json` loader hash what it loads, assert that it matches the expected value, and only proceed if it matches. "No evaluation of unverified code". At each point we're about to load a file, we should know the expected hash ahead of time, and we don't convert the inert bytes into behavior until after it's passed the check.
* It sounds like the contents of the compartment map determines whether or not a hash is checked on each module. That sounds reasonable, but we must make sure that this doesn't enable an attacker to substitute a doctored map (lacking module hashes) to disable any checks. I think the hash check of the map itself is sufficient to prevent this (they can't remove the module hashes and still meet the expected map hash), but "in-band" security markers that have broken many others systems (the JWT `"alg": "none"` field comes to mind).

Reply:
* This is a good catch and I’m revising the `importArchive`, `loadArchive`, and `parseArchive` functions to accept `expectedSha512` which will get checked against the archived `compartment-map.json`.
* If you pass `computeSha512`, every module must have a `sha512` and pass an integrity check, regardless of whether you also provide `expectedSha512`.